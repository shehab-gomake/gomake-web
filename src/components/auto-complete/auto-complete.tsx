import * as React from "react";
import TextField from "@mui/material/TextField";
import { styled } from "@mui/material/styles";
import Autocomplete from "@mui/material/Autocomplete";
import { ArrowDownIcon } from "@/icons/arrow-down";

const StyledAutocomplete: any = styled(Autocomplete)((props: any) => ({
  boxSizing: "border-box",
  borderRadius: "4px",
  height: props?.style?.height || 40,
  fontFamily: "Lexend",
  fontStyle: "normal",
  fontWeight: 300,
  fontSize: 14,
  lineHeight: "21px",
  display: "flex",
  alignItems: "center",
  color: props?.error ? "red" : "#B9B9D9",
  border: props?.error ? "1px solid red" : "1px solid #9E9E9E",
  "& .MuiOutlinedInput-root": {
    color: props?.error ? "red" : "",
    height: props?.style?.height || 40,
    fontFamily: "Lexend",
    fontStyle: "normal",
    fontWeight: 300,
    "& fieldset": {
      border: "transparent",
      boxSizing: "border-box",
      borderRadius: "4px",
    },
    "& .MuiAutocomplete-input": {
      padding: 0,
    },
  },
  "& .MuiAutocomplete-endAdornment": {
    top: "calc(50% - 12px)",
  },
}));

const GoMakeAutoComplate = ({
  value,
  onChange,
  style,
  error,
  options,
  autoHighlight,
  getOptionLabel,
  renderOption,
  disableClearable,
  placeholder,
  defaultValue,
  multiple = false,
}: {
  value?: string;
  onChange?: any;
  style?: any;
  error?: any;
  options: any;
  autoHighlight?: any;
  getOptionLabel?: any;
  renderOption?: any;
  disableClearable?: any;
  placeholder?: any;
  defaultValue?: any;
  multiple?: any;
}) => {
  return (
    <StyledAutocomplete
      value={value}
      onChange={onChange}
      style={style}
      options={options}
      popupIcon={<ArrowDownIcon />}
      renderInput={(params: any) => (
        <TextField
          {...params}
          placeholder={!multiple && (defaultValue?.name || placeholder)}
        />
      )}
      autoHighlight={autoHighlight}
      getOptionLabel={getOptionLabel}
      renderOption={renderOption}
      error={error}
      disableClearable={disableClearable}
      placeholder="Enter"
      multiple={multiple}
      isOptionEqualToValue={(option: any, value: any) => {
        return option.value === value.value;
      }}
    />
  );
};

export { GoMakeAutoComplate };
