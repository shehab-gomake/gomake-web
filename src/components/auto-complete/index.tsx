export * from "./auto-complete";
export * from "./update-value-auto-complete";
