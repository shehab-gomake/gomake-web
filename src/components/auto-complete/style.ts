import { useGomakeTheme } from "@/hooks/use-gomake-thme";
import { useMemo } from "react";
import { FONT_FAMILY } from "@/utils/font-family";

const useStyle = () => {
    const classes = useMemo(() => {
        return {
            multiSelectStyle: {
                display:"flex",
                width:"200px",
                border: "0px"
            },
            multiSelectOption:{
                width: "100%",
                ...FONT_FAMILY.Lexend(400, 14),
                display: "flex",
                flexDirection : "row" as "row",
                alignItems:'center'
              },

        };
    }, []);
    return {
        classes,
    };
};
export { useStyle };