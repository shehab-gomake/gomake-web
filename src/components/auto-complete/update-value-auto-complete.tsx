import { useGomakeTheme } from "@/hooks/use-gomake-thme";
import { Autocomplete, Paper, TextField } from "@mui/material";
import { FONT_FAMILY } from "@/utils/font-family";
import { ClickOutside } from "@/components/click-out-side/click-out-side";
import { RouteChangeConfirmation } from "@/components/handle-navigation/handle-navigation";
import { ArrowUpIcon } from "@/icons";

interface IUpdateValueAutoCompleteProps {
  value?: any;
  onUpdate?: () => void;
  onCancel?: () => void;
  onInputChange?: (e: string) => void;
  clickedOut?: () => void;
  options?: any;
  onChange?: any;
  getOptionLabel?: any;
  onChangeTextField?: any;
}

const UpdateValueAutoComplete = ({
  clickedOut,
  options,
  onChange,
  getOptionLabel,
  onChangeTextField
}: IUpdateValueAutoCompleteProps) => {
  const { secondColor } = useGomakeTheme();
  return (
    <>
      <ClickOutside onClick={clickedOut}>
        <RouteChangeConfirmation />
        <Paper
          component="form"
          sx={{
            p: "4px 4px",
            display: "inline-flex",
            alignItems: "center",
            width: 137,
            height: "26px",
            border: "1px solid " + secondColor(400),
            boxShadow: "none",
            position: "relative",
          }}
        >
          <Autocomplete
            id="updateValueAutoComplete"
            options={options}
            sx={{
              ml: 1,
              flex: 1,
              ...FONT_FAMILY.Lexend(400, 14),
              padding: 0,
              position: "relative",
            }}
            getOptionLabel={getOptionLabel}
            renderInput={(params) => <TextField onChange={onChangeTextField} {...params} />}
            onChange={((e: any, value: any) => {
              onChange(e, value);
            })}
            popupIcon={
              <div style={{ position: "absolute", top: -11 }}>
                <ArrowUpIcon />
              </div>
            }
            style={{ height: 26 }}
            clearIcon={false}
          />
        </Paper>
      </ClickOutside>
    </>
  );
};

export { UpdateValueAutoComplete };