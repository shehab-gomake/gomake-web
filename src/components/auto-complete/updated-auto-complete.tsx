import React, { useState, useEffect } from "react";
import { Autocomplete, Paper, TextField, IconButton, debounce } from "@mui/material";
import CloseIcon from "@mui/icons-material/Close";
import CheckIcon from "@mui/icons-material/Check";
import { ClickOutside } from "@/components/click-out-side/click-out-side";
import { RouteChangeConfirmation } from "@/components/handle-navigation/handle-navigation";
import { ArrowUpIcon } from "@/icons";
import { useRef } from "react";

interface IUpdateValueAutoCompleteProps {
    value?: any;
    onUpdate?: () => void;
    onCancel?: () => void;
    onInputChange?: (e: string) => void;
    clickedOut?: () => void;
    options?: any;
    onChange?: any;
    onChangeTextField?: any;
    height?: string;
    width?: string;
}

const UpdatedAutoComplete = ({
    onUpdate,
    onCancel,
    value,
    clickedOut,
    options,
    onChange,
    onChangeTextField,
    width,
    height
}: IUpdateValueAutoCompleteProps) => {
    const optionsListEl = useRef(null);
    const [isFocused, setIsFocused] = useState(false);
    const [originalValue, setOriginalValue] = useState<string>(value);
    const [updatedValue, setUpdatedValue] = useState<string>(value);

    useEffect(() => {
        setOriginalValue(value);
        setUpdatedValue(value);
    }, [value]);

    const handleFocus = () => {
        setIsFocused(true);
    };

    const handleBlur = () => {
        setIsFocused(false);
        // onCancel();
    };


    const handleCancelClick = () => {
        onCancel();
    };

    const handleUpdateClick = () => {
        onUpdate();
    };

    // const handleChange = (event, newValue) => {
    //     setUpdatedValue(newValue);
    //     if (onChange) {
    //         onChange(event, newValue);
    //     }
    // };

    const handleChange = (event, newValue) => {
        setUpdatedValue(newValue);
        const originalNumberValue = parseFloat(originalValue);
        if (newValue !== originalNumberValue) {
            onUpdate();
        }
        if (onChange) {
            onChange(event, newValue);
        }
    };
    return (
        <>
            <Paper
                component="form"
                sx={{
                    p: "4px 4px",
                    display: "inline-flex",
                    alignItems: "center",
                    width: width || "137px",
                    height: height || "26px",
                    borderRadius: "5px",
                    boxShadow: "0px 1px 2px 0px rgba(0, 0, 0, 0.08)",
                    backgroundColor: "#FFF",
                }}
                onBlur={handleBlur}
            >
                <Autocomplete
                    id="updateValueAutoComplete"
                    options={options}
                    value={updatedValue}
                    sx={{
                        ml: 1,
                        flex: 1,
                        padding: 0,
                        position: "relative",
                    }}
                    renderInput={(params) => (
                        <TextField
                            {...params}
                            onChange={onChangeTextField}
                            onFocus={handleFocus}
                            autoFocus={isFocused}
                        />
                    )}
                    onChange={handleChange}
                    popupIcon={
                        <div style={{ position: "absolute", top: "50%", transform: "translateY(-50%)" }}>
                            <ArrowUpIcon />
                        </div>
                    }
                    clearIcon={false}
                    PaperComponent={(props) => (
                        <Paper ref={optionsListEl} {...props}>
                            {props?.children}
                        </Paper>
                    )}
                />
                {/* {isFocused && (
                    <div>
                        <IconButton
                            id="cancel-icon"
                            onClick={handleCancelClick}
                            sx={{
                                m: "0 4px",
                                backgroundColor: "#FFF",
                                borderRadius: "5px",
                                color: secondColor(600),
                                height: "16px",
                                width: "16px",
                            }}
                        >
                            <CloseIcon style={{ width: "12px", height: "12px" }} />
                        </IconButton>
                        <IconButton
                            id="update-icon"
                            onClick={handleUpdateClick}
                            sx={{
                                backgroundColor: "#FFF",
                                borderRadius: "5px",
                                color: secondColor(600),
                                height: "16px",
                                width: "16px",
                            }}
                        >
                            <CheckIcon style={{ width: "12px", height: "12px" }} />
                        </IconButton>
                    </div>
                )} */}
            </Paper>
        </>
    );
};

export { UpdatedAutoComplete };