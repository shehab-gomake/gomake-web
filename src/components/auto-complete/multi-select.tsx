import * as React from "react";
import { Checkbox, Paper } from "@mui/material";
import { CheckboxCheckedIcon, CheckboxIcon } from "@/icons";
import Stack from "@mui/material/Stack";
import { GoMakeAutoComplate } from "./auto-complete";
import { useStyle } from "./style";

interface IProps {
    options: Array<{ value: string; label: string }>;
    values: string[];
    onChange: (values: string[]) => void;
    style?: React.CSSProperties;
    placeholder?: string;
}

const GoMakeMultiSelect = ({ options, values, style, onChange , placeholder="" }: IProps) => {
    const { classes } = useStyle();
    const reorderedOptions = options.sort((a, b) => {
        const isSelectedA = values.includes(a.value);
        const isSelectedB = values.includes(b.value);

        if (isSelectedA && !isSelectedB) {
            return -1;
        } else if (!isSelectedA && isSelectedB) {
            return 1;
        } else {
            return 0;
        }
    });

    const selectedLabels = values.map(selectedId => options.find(opt => opt.value === selectedId)?.label)
        .filter(label => label !== null && label !== undefined);

        const handleSelectCheck = (isChecked: boolean, option: any) => {
            const updatedValues = isChecked
              ? [...values, option?.value]
              : values.filter(v => v !== option?.value);
            onChange && onChange(updatedValues);
          };

    return (
        <GoMakeAutoComplate
            style={style || classes.multiSelectStyle}
            value={selectedLabels}
            disableClearable={true} placeholder={placeholder}
            options={reorderedOptions}
            renderOption={(props: any, option: any) => {
                return (
                    <Stack style={classes.multiSelectOption}>
                        <div>
                            <Checkbox
                                onChange={(e, checked) => handleSelectCheck(checked, option)}
                                icon={<CheckboxIcon />}
                                checkedIcon={<CheckboxCheckedIcon />}
                                checked={values?.includes(option?.value)} />
                        </div>
                        <div>
                            {option.label}
                        </div>
                    </Stack>
                )
            }}
            PaperComponent={(props: any) => <Paper style={style || classes.multiSelectStyle} {...props}>{props?.children}</Paper>}
        />
    );
};

export { GoMakeMultiSelect };
