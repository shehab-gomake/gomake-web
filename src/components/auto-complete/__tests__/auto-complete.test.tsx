import renderer from "react-test-renderer";
import { GoMakeAutoComplate } from "../auto-complete";

it("renders correctly", () => {
  const tree = renderer.create(<GoMakeAutoComplate options={[]} />).toJSON();
  expect(tree).toMatchSnapshot();
});
