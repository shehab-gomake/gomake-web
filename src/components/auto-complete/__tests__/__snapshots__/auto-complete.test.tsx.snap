// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-mz9wi4-MuiAutocomplete-root"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  placeholder="Enter"
>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-expanded={false}
        aria-invalid={false}
        autoCapitalize="none"
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        readOnly={false}
        required={false}
        role="combobox"
        spellCheck="false"
        type="text"
        value=""
      />
      <div
        className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
      >
        <button
          aria-label="Open"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Open"
          type="button"
        >
          <svg
            fill="none"
            height="11"
            viewBox="0 0 20 11"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0.458303 1.525C0.11525 1.19167 0.0951939 0.854333 0.398138 0.513C0.701082 0.171666 1.16975 0.000666667 1.80414 0L18.1916 0C18.825 0 19.2936 0.171 19.5976 0.513C19.9016 0.855 19.8816 1.19233 19.5375 1.525L11.3437 9.675C11.0271 9.975 10.5784 10.125 9.99788 10.125C9.41733 10.125 8.96872 9.975 8.65205 9.675L0.458303 1.525Z"
              fill="#F135A3"
            />
          </svg>
        </button>
      </div>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-ihdtdm"
        >
          <span
            className="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
